internal/controller/multiclusterservice_controller.go

```diff
@@ -29,10 +37,132 @@ type MultiClusterServiceReconciler struct {
 }
 
 // Reconcile reconciles a MultiClusterService object.
-func (*MultiClusterServiceReconciler) Reconcile(ctx context.Context, _ ctrl.Request) (ctrl.Result, error) {
-	_ = ctrl.LoggerFrom(ctx)
+func (r *MultiClusterServiceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	l := ctrl.LoggerFrom(ctx).WithValues("MultiClusterServiceController", req.NamespacedName.String())
+	l.Info("Reconciling MultiClusterService")
 
-	// TODO(https://github.com/Mirantis/hmc/issues/455): Implement me.
+	mcsvc := &hmc.MultiClusterService{}
+	err := r.Get(ctx, req.NamespacedName, mcsvc)
+	if apierrors.IsNotFound(err) {
+		l.Info("MultiClusterService not found, ignoring since object must be deleted")
+		return ctrl.Result{}, nil
+	}
+	if err != nil {
+		l.Error(err, "Failed to get MultiClusterService")
+		return ctrl.Result{}, err
+	}
+
+	if !mcsvc.DeletionTimestamp.IsZero() {
+		l.Info("Deleting MultiClusterService")
+		return r.reconcileDelete(ctx, mcsvc)
+	}
+
+	if ok := controllerutil.AddFinalizer(mcsvc, hmc.MultiClusterServiceFinalizer); ok {
+		if err := r.Client.Update(ctx, mcsvc); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to update MultiClusterService %s with finalizer %s: %w", mcsvc.Name, hmc.MultiClusterServiceFinalizer, err)
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// By using DefaultSystemNamespace we are enforcing that MultiClusterService
+	// may only use ServiceTemplates that are present in the hmc-system namespace.
+	opts, err := HelmChartOpts(ctx, r.Client, l, utils.DefaultSystemNamespace, mcsvc.Spec.Services)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if _, err := sveltos.ReconcileClusterProfile(ctx, r.Client, l, mcsvc.Name,
+		sveltos.ReconcileProfileOpts{
+			OwnerReference: &metav1.OwnerReference{
+				APIVersion: hmc.GroupVersion.String(),
+				Kind:       hmc.MultiClusterServiceKind,
+				Name:       mcsvc.Name,
+				UID:        mcsvc.UID,
+			},
+			LabelSelector:  mcsvc.Spec.ClusterSelector,
+			HelmChartOpts:  opts,
+			Priority:       mcsvc.Spec.Priority,
+			StopOnConflict: mcsvc.Spec.StopOnConflict,
+		}); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to reconcile ClusterProfile: %w", err)
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// HelmChartOpts returns slice of helm chart options to use with Sveltos.
+// Namespace is the namespace of the referred templates in services slice.
+func HelmChartOpts(ctx context.Context, c client.Client, l logr.Logger, namespace string, services []hmc.ServiceSpec) ([]sveltos.HelmChartOpts, error) {
```

Comment:
- It doesn't need to be public (can change that)
- I passed logger so that it will have the WithValues() already set from where this function is called e.g. [here](https://github.com/wahabmk/hmc/blob/07392894f35ba7d2a37b2140f15d0487931f0d25/internal/controller/managedcluster_controller.go#L374-L377).
- I named it without `get*` because [I had this](https://go.dev/doc/effective_go#Getters) in mind. However, maybe this doesn't qualify as a getter? In any case I am open to change it to `getHelmChartOpts` if that is preferable.

Reply:
regarding the name: thanks, haven't had it in mind, then it's ok
regarding the logger: there is `ctrl.LoggerInto(ctx)` which provides convenient way to change kind-of-global logger and put it back in the context, but having this logger there is no need, since those values are already embedded into the default logger so there is no need at all to duplicate this info (that's why I remove such an approach elsewhere)